libs/ml/include/ml/graph.hpp

```diff
@@ -218,28 +223,69 @@ void Graph<ArrayType>::ApplyRegularisation()
  */
 template <typename ArrayType>
 template <class OperationType, typename... Params>
-std::string Graph<ArrayType>::AddNode(std::string const &             node_name,
-                                      std::vector<std::string> const &inputs, Params... params)
+meta::IfIsShareable<ArrayType, OperationType, std::string> Graph<ArrayType>::AddNode(
+    std::string const &node_name, std::vector<std::string> const &inputs, Params... params)
 {
   // guarantee unique op name
-  std::string name = UpdateVariableName<OperationType>(node_name);
+  auto        naming_result = UpdateVariableName<OperationType>(node_name);
```

Comment:
don't create and return a pair with UpdateVariableName; instead return only a bool, and change the input string const & to a string & then update the name internally.

Then, in AddNode you can just do:
bool is_duplicate = UpdateVariableName<OperationType>(node_name);

(and then re-use node_name later instead of "updated_name")